[site-isolation] window.open should work with blob urls for file URL's

jelee53 opened the pull request:
#### adc30c72f52200f1173b96d28fb135c4de3c4ce5
<pre>
[site-isolation] window.open should work with blob urls
<a href="https://bugs.webkit.org/show_bug.cgi?id=302702">https://bugs.webkit.org/show_bug.cgi?id=302702</a>
<a href="https://rdar.apple.com/136853662">rdar://136853662</a>

Reviewed by Chris Dumez.

This case only fixes the edge case in which the mainframe is a file URL.

In this change, we dedicate a new process for cross-site blob URL&apos;s.
We also forward the top origin of the mainframe requester to the new process
dedicated to the blob URL.

Tests: Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj
       Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/blob-popup-file-mainframe.html
       Tools/TestWebKitAPI/Tests/WebKitCocoa/blob-popup-local-iframe.html

* Source/WebCore/loader/DocumentLoader.cpp:
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::crossSiteRequester const):
(WebCore::DocumentLoader::setCrossSiteRequester):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::load):
(WebCore::FrameLoader::loadWithDocumentLoader):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/NavigationAction.cpp:
(WebCore::shouldTreatAsSameOriginNavigation):
(WebCore::NavigationAction::NavigationAction):
* Source/WebCore/loader/NavigationAction.h:
* Source/WebCore/loader/NavigationRequester.h:
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::requesterOrigin const):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::requesterOrigin const): Deleted.
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::interceptMarketplaceKitNavigation):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/win/AutomationSessionClientWin.cpp:
(WebKit::AutomationSessionClient::requestNewPageWithOptions):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJS):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):
* Source/WebKitLegacy/ios/WebCoreSupport/WebFixedPositionContent.mm:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, LocalIframeOpensBlobURLFromFileMainFrame)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/blob-popup-file-mainframe.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/blob-popup-local-iframe.html: Added.

Canonical link: <a href="https://commits.webkit.org/303469@main">https://commits.webkit.org/303469@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/284e1dde9feb9bf964ac7e58e94d8b928a4832de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132544 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5039 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43604 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140063 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84548 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0d11c12f-f620-41d0-8a07-b0e6fff37697) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134414 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/5197 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4798 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/101331 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/917186e6-bc13-4164-8275-27df3267b232) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135490 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/118731 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82126 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d5f053b8-8919-43d5-88b5-8694e14c49a0) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1324 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83298 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112451 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36848 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142717 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4709 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37436 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/109707 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4791 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4070 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109889 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3581 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115004 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58130 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20574 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4763 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33354 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4598 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68214 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4854 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4720 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jelee53 commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -544,6 +545,9 @@ class DocumentLoader
     WEBCORE_EXPORT void setNewResultingClientId(ScriptExecutionContextIdentifier);
 
+    std::optional<NavigationRequester> crossSiteRequester() const { return m_crossSiteRequester; }
+    void setCrossSiteRequester(NavigationRequester& crossSiteRequester) { m_crossSiteRequester = crossSiteRequester; }
```
@pvollan @szewai does the `crossSiteRequester ` argument need to be `const` or can I leave as-is?

pvollan replied:
I think it would be good to add `const` here.

pvollan commented on Source/WebCore/loader/EmptyClients.cpp:
```diff
@@ -615,6 +615,12 @@ RefPtr<SearchPopupMenu> EmptyChromeClient::createSearchPopupMenu(PopupMenuClient
     return adoptRef(*new EmptySearchPopupMenu);
 }
 
+RefPtr<Page> EmptyChromeClient::createWindow(LocalFrame&, const String&, const WindowFeatures&, const NavigationAction&, const std::optional<Ref<Document>> requester)
```
No need for `const` in the `requester` parameter, since we're passing this by value.

cdumez commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -544,6 +545,9 @@ class DocumentLoader
 
     WEBCORE_EXPORT void setNewResultingClientId(ScriptExecutionContextIdentifier);
 
+    std::optional<NavigationRequester> crossSiteRequester() const { return m_crossSiteRequester; }
```
Ideally would return `const std::optional<NavigationRequester>&` to avoid copying.

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1713,7 +1713,7 @@ SubstituteData FrameLoader::defaultSubstituteDataForURL(const URL& url)
 }
 
-void FrameLoader::load(FrameLoadRequest&& request)
+void FrameLoader::load(FrameLoadRequest&& request, std::optional<NavigationRequester> crossSiteRequester)
```
We should not be passing `std::optional<NavigationRequester>` by value. Either r-value reference or const reference. Here I think the caller can move the value in so I'd use an r-value reference for ideal performance.

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1751,6 +1751,9 @@ void FrameLoader::load(FrameLoadRequest&& request)
     loader->setIsRequestFromClientOrUserInput(request.isRequestFromClientOrUserInput());
     loader->setIsContinuingLoadAfterProvisionalLoadStarted(request.shouldTreatAsContinuingLoad() == ShouldTreatAsContinuingLoad::YesAfterProvisionalLoadStarted);
+    if (crossSiteRequester)
+        loader->setCrossSiteRequester(*crossSiteRequester);
```
if you manage to take in a r-value reference, please use `WTFMove(*crossSiteRequester)` here and update `setCrossSiteRequester` to take in a r-value reference and use `WTFMove()` internally.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -2200,7 +2200,7 @@ void WebPage::loadRequest(LoadParameters&& loadParameters)
     localFrame->loader().setHTTPFallbackInProgress(loadParameters.isPerformingHTTPFallback);
     localFrame->loader().setRequiredCookiesVersion(loadParameters.requiredCookiesVersion);
-    localFrame->loader().load(WTFMove(frameLoadRequest));
+    localFrame->loader().load(WTFMove(frameLoadRequest), loadParameters.requester);
```
You can use `WTFMove(loadParameters.requester)` here which I think means you can move the request all the way in to the DocumentLoader to avoid any copying.

cdumez commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -544,6 +545,9 @@ class DocumentLoader
     WEBCORE_EXPORT void setNewResultingClientId(ScriptExecutionContextIdentifier);
 
+    std::optional<NavigationRequester> crossSiteRequester() const { return m_crossSiteRequester; }
+    void setCrossSiteRequester(const NavigationRequester& crossSiteRequester) { m_crossSiteRequester = crossSiteRequester; }
```
`NavigationRequester&&` and `WTFMove()` internally since the call site can move the requester in I believe.

cdumez commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -771,6 +775,8 @@ class DocumentLoader
     PushAndNotificationsEnabledPolicy m_pushAndNotificationsEnabledPolicy { PushAndNotificationsEnabledPolicy::UseGlobalPolicy };
     InlineMediaPlaybackPolicy m_inlineMediaPlaybackPolicy { InlineMediaPlaybackPolicy::Default };
     WebpagePreferences m_preferences;
+    // The triggering action's requester should take precendence. This is used for site-isolation situations that require a cross-site requester.
```
typo: `precendence`

cdumez commented on Source/WebCore/loader/EmptyClients.h:
```diff
@@ -74,7 +74,7 @@ class EmptyChromeClient : public ChromeClient {
     void focusedFrameChanged(Frame*) final { }
 
-    RefPtr<Page> createWindow(LocalFrame&, const String&, const WindowFeatures&, const NavigationAction&) final { return nullptr; }
+    RefPtr<Page> createWindow(LocalFrame&, const String&, const WindowFeatures&, const NavigationAction&, std::optional<Ref<Document>> requester) final;
```
`std::optional<Ref<Document>>` doesn't make much sense, I think what you mean is `RefPtr<Document>`. But also, I doubt this needs to be a smart pointer at all, maybe just `Document*`

cdumez commented on Source/WebCore/loader/NavigationAction.cpp:
```diff
@@ -131,6 +136,21 @@ NavigationAction::NavigationAction(Document& requester, const ResourceRequest& o
     setIsRequestFromClientOrUserInput(isRequestFromClientOrUserInput);
 }
 
+NavigationAction::NavigationAction(const NavigationRequester& requester, const ResourceRequest& originalRequest, InitiatedByMainFrame initiatedByMainFrame, bool isRequestFromClientOrUserInput, FrameLoadType frameLoadType, bool isFormSubmission, Event* event, ShouldOpenExternalURLsPolicy shouldOpenExternalURLsPolicy, const AtomString& downloadAttribute, Element* sourceElement)
```
Can we have the constructor above call your new constructor to avoid duplication?
e.g.
```
NavigationAction::NavigationAction(Document& requester, const ResourceRequest& originalRequest, InitiatedByMainFrame initiatedByMainFrame, bool isRequestFromClientOrUserInput, FrameLoadType frameLoadType, bool isFormSubmission, Event* event, ShouldOpenExternalURLsPolicy shouldOpenExternalURLsPolicy, const AtomString& downloadAttribute, Element* sourceElement)
    : NavigationAction(NavigationRequester::from(requester), originalRequest, /* ... */)
{ }
```

cdumez commented on Source/WebCore/loader/NavigationAction.h:
```diff
@@ -60,7 +60,9 @@ class NavigationAction : public FrameLoadRequestBase {
     WEBCORE_EXPORT NavigationAction(Document&, const ResourceRequest&, InitiatedByMainFrame, bool, NavigationType = NavigationType::Other, ShouldOpenExternalURLsPolicy = ShouldOpenExternalURLsPolicy::ShouldNotAllow, Event* = nullptr, const AtomString& downloadAttribute = nullAtom(), Element* sourceElement = nullptr);
+    WEBCORE_EXPORT NavigationAction(NavigationRequester&, const ResourceRequest&, InitiatedByMainFrame, bool, NavigationType = NavigationType::Other, ShouldOpenExternalURLsPolicy = ShouldOpenExternalURLsPolicy::ShouldNotAllow, Event* = nullptr, const AtomString& downloadAttribute = nullAtom(), Element* sourceElement = nullptr);
     NavigationAction(Document&, const ResourceRequest&, InitiatedByMainFrame, bool, FrameLoadType, bool isFormSubmission, Event* = nullptr, ShouldOpenExternalURLsPolicy = ShouldOpenExternalURLsPolicy::ShouldNotAllow, const AtomString& downloadAttribute = nullAtom(), Element* sourceElement = nullptr);
+    NavigationAction(const NavigationRequester&, const ResourceRequest&, InitiatedByMainFrame, bool, FrameLoadType, bool isFormSubmission, Event* = nullptr, ShouldOpenExternalURLsPolicy = ShouldOpenExternalURLsPolicy::ShouldNotAllow, const AtomString& downloadAttribute = nullAtom(), Element* sourceElement = nullptr);
```
Why this constructor in addition to the other one? Please limit code changes to the minimum. I bet we don't need 2 constructors though I did not look much into it.

cdumez commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -241,7 +241,7 @@ class ChromeClient {
     // created Page has its show method called.
     // The ChromeClient should not load the request.
-    virtual RefPtr<Page> createWindow(LocalFrame&, const String& openedMainFrameName, const WindowFeatures&, const NavigationAction&) = 0;
+    virtual RefPtr<Page> createWindow(LocalFrame&, const String& openedMainFrameName, const WindowFeatures&, const NavigationAction&, std::optional<Ref<Document>> requester = std::nullopt) = 0;
```
Again, this is the wrong type everywhere. `std::optional<Ref<T>>` is not really a thing in WebKit

cdumez commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -149,6 +149,13 @@ void Navigation::markRequestAsFromClientInput()
+WebCore::SecurityOriginData Navigation::requesterOrigin() const
+{
+    if (m_lastNavigationAction && m_lastNavigationAction->requester)
+        return (*m_lastNavigationAction->requester).securityOrigin->data();
```
Can we use `->` instead of `(*foo)`?

cdumez commented on Source/WebKit/UIProcess/API/APINavigation.cpp:
```diff
@@ -149,6 +149,13 @@ void Navigation::markRequestAsFromClientInput()
+{
+    if (m_lastNavigationAction && m_lastNavigationAction->requester)
+        return (*m_lastNavigationAction->requester).securityOrigin->data();
+    return WebCore::SecurityOriginData { };
```
Probably `return { };` is enough.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2189,6 +2189,8 @@ void WebPageProxy::loadRequestWithNavigationShared(Ref<WebProcessProxy>&& proces
     loadParameters.isHandledByAboutSchemeHandler = m_aboutSchemeHandler->canHandleURL(url);
     loadParameters.requiredCookiesVersion = protectedWebsiteDataStore()->cookiesVersion();
     loadParameters.originatingFrame = navigation.lastNavigationAction() ? std::optional(navigation.lastNavigationAction()->originatingFrameInfoData) : std::nullopt;
+    if (navigation.lastNavigationAction())
```
Normally we'd do:
```
if (auto& action = navigation.lastNavigationAction())
    loadParameters.requester = action->requester;
```

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5171,6 +5173,9 @@ void WebPageProxy::receivedNavigationActionPolicyDecision(WebProcessProxy& proce
             loadParameters.ownerPermissionsPolicy = navigation->ownerPermissionsPolicy();
             loadParameters.isPerformingHTTPFallback = navigationAction->data().isPerformingHTTPFallback;
             loadParameters.isHandledByAboutSchemeHandler = m_aboutSchemeHandler->canHandleURL(loadParameters.request.url());
+            if (navigation->lastNavigationAction())
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5442,6 +5447,8 @@ void WebPageProxy::continueNavigationInNewProcess(API::Navigation& navigation, W
         loadParameters.ownerPermissionsPolicy = navigation.ownerPermissionsPolicy();
         loadParameters.isPerformingHTTPFallback = isPerformingHTTPFallback == IsPerformingHTTPFallback::Yes;
         loadParameters.isHandledByAboutSchemeHandler = m_aboutSchemeHandler->canHandleURL(loadParameters.request.url());
+        if (navigation.lastNavigationAction())
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9025,7 +9032,8 @@ void WebPageProxy::createNewPage(IPC::Connection& connection, WindowFeatures&& w
             newPage->internals().privateClickMeasurement = { { WTFMove(*privateClickMeasurement), { }, { } } };
 
-        if (navigationDataForNewProcess && !openedBlobURL) {
+        bool isBlobURLOpenedInSiteIsolation = openedBlobURL && protectedPreferences()->siteIsolationEnabled();
```
Checking `siteIsolationEnabled()` here is a little fishy. This may be correct but this at least deserves a code comment to explain the reasoning.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -410,6 +408,7 @@ RefPtr<Page> WebChromeClient::createWindow(LocalFrame& frame, const String& open
         originalRequest, /* originalRequest */
         originalRequest, /* request */
         originalRequest.url().isValid() ? String() : originalRequest.url().string(), /* invalidURLString */
+        requester ? NavigationRequester::from(requester.value().get()) : std::optional<NavigationRequester>(std::nullopt), /* requester */
```
Don't need `std::nullopt`

cdumez commented on Source/WebCore/loader/DocumentLoader.h:
```diff
@@ -544,6 +545,9 @@ class DocumentLoader
 
     WEBCORE_EXPORT void setNewResultingClientId(ScriptExecutionContextIdentifier);
 
+    const std::optional<NavigationRequester> crossSiteRequester() const { return m_crossSiteRequester; }
```
Should return a `const std::optional<NavigationRequester>&`

jelee53 replied:
Oops, I'm so sorry for forgetting to make this change! I did see your feedback earlier about this. Thank you for catching this and for approving. Apologies again for not fixing this earlier, and I've made these changes. 

I also forgot to re-add the test files when I pushed these changes. I've re-added them so hopefully it will build now.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -410,6 +408,7 @@ RefPtr<Page> WebChromeClient::createWindow(LocalFrame& frame, const String& open
         originalRequest, /* originalRequest */
         originalRequest, /* request */
         originalRequest.url().isValid() ? String() : originalRequest.url().string(), /* invalidURLString */
+        requester ? NavigationRequester::from(*requester) : std::optional<NavigationRequester>(), /* requester */
```
You can get the requester from `navigationAction.requester()` so no need to have the `requester` parameter for the `createWindow` functions

jelee53 replied:
I think that when I had looked at that originally, navigationAction.requester showed as `about:blank` which I didn't want. I will double check this again.

szewai commented on Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
```diff
@@ -463,8 +463,9 @@ static void interceptMarketplaceKitNavigation(Ref<API::NavigationAction>&& actio
-    if (!action->shouldOpenExternalSchemes() || !action->isProcessingUserGesture() || action->isRedirect() || action->data().requesterTopOrigin.isNull()) {
-        RELEASE_LOG_ERROR(Loading, "NavigationState: can't handle MarketplaceKit navigation with shouldOpenExternalSchemes: %d, isProcessingUserGesture: %d, isRedirect: %d, requesterTopOriginIsNull: %d", action->shouldOpenExternalSchemes(), action->isProcessingUserGesture(), action->isRedirect(), action->data().requesterTopOrigin.isNull());
+    auto requesterTopOrigin = action->data().requester ? std::optional { action->data().requester->topOrigin } : std::nullopt;
+    if (!action->shouldOpenExternalSchemes() || !action->isProcessingUserGesture() || action->isRedirect() || requesterTopOrigin) {
```
The original condition is `requesterTopOrigin.isNull()`, the patch changes it to `requesterTopOrigin`

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9025,7 +9032,11 @@ void WebPageProxy::createNewPage(IPC::Connection& connection, WindowFeatures&& w
+        bool isCrossSiteNavigation = Site(mainFrameURL) != Site(request.url());
+        bool blobRequiresNewProcess = openedBlobURL && protectedPreferences()->siteIsolationEnabled() && isCrossSiteNavigation;
+
+        if (navigationDataForNewProcess && (blobRequiresNewProcess || !openedBlobURL)) {
```
Why not just `if (navigationDataForNewProcess && (protectedPreferences()->siteIsolationEnabled() || !openedBlobURL)) {`?

szewai commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -7235,6 +7243,7 @@
 					};
 					A17C46452C98E3430023F3C7 = {
 						CreatedOnToolsVersion = 16.0;
+						LastSwiftMigration = 2600;
```
Remove the irrelevant changes

szewai commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -8600,13 +8609,18 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = A17C464A2C98E3640023F3C7 /* TestWebKitAPIResources.xcconfig */;
 			buildSettings = {
+				CLANG_ENABLE_MODULES = YES;
```
Remove the irrelevant changes

szewai commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -8600,13 +8609,18 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = A17C464A2C98E3640023F3C7 /* TestWebKitAPIResources.xcconfig */;
 			buildSettings = {
+				CLANG_ENABLE_MODULES = YES;
```
Remove the irrelevant changes

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -6833,6 +6833,63 @@ HTTPServer server({
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600)]);
+
+    HTTPServer server({
+        { "/blob-popup-local-iframe.html"_s, { [NSData dataWithContentsOfURL:[NSBundle.test_resourcesBundle URLForResource:@"blob-popup-local-iframe" withExtension:@"html"]] } }
```
You don't need to add a file for `iframe.html`, just do:
```
auto iframeHTML = "<script>...</script>"_s;
HTTPServer server({
    { "/blob-popup-local-iframe.html"_s, { iframeHTML } },
}, HTTPServer::Protocol::HttpsProxy);
```

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -6833,6 +6833,63 @@ HTTPServer server({
+    [webView loadFileURL:file allowingReadAccessToURL:file.URLByDeletingLastPathComponent];
+
+    // Add a delay so that the main frame has sufficient time to receive the message
+    TestWebKitAPI::Util::runFor(2_s);
```
We don't normally rely on delay unless we don't have a way to know when the result is ready.
In this case you could just let the main frame html post message back to test runner, see the example of "window.webkit.messageHandlers.testHandler.postMessage" in other tests.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -6833,6 +6833,63 @@ HTTPServer server({
+
+    if (blobWindow) {
+        __block bool blobContentChecked = false;
+        [blobWindow evaluateJavaScript:@"document.body.innerHTML" completionHandler:^(id result, NSError *error) {
```
Same here, you could add `alert(...)` or `window.webkit.messageHandlers.testHandler.postMessage(...)` in the blob page's javascript, and then catch wait for the callback with UIDelegate.

jelee53 committed:
[site-isolation] window.open should work with blob urls
https://bugs.webkit.org/show_bug.cgi?id=302702
rdar://136853662

Reviewed by Chris Dumez.

This case only fixes the edge case in which the mainframe is a file URL.

In this change, we dedicate a new process for cross-site blob URL's.
We also forward the top origin of the mainframe requester to the new process
dedicated to the blob URL.

Tests: Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj
       Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
       Tools/TestWebKitAPI/Tests/WebKitCocoa/blob-popup-file-mainframe.html
       Tools/TestWebKitAPI/Tests/WebKitCocoa/blob-popup-local-iframe.html

* Source/WebCore/loader/DocumentLoader.cpp:
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::crossSiteRequester const):
(WebCore::DocumentLoader::setCrossSiteRequester):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::load):
(WebCore::FrameLoader::loadWithDocumentLoader):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/NavigationAction.cpp:
(WebCore::shouldTreatAsSameOriginNavigation):
(WebCore::NavigationAction::NavigationAction):
* Source/WebCore/loader/NavigationAction.h:
* Source/WebCore/loader/NavigationRequester.h:
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/LoadParameters.serialization.in:
* Source/WebKit/Shared/NavigationActionData.h:
* Source/WebKit/Shared/NavigationActionData.serialization.in:
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::requesterOrigin const):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::requesterOrigin const): Deleted.
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::interceptMarketplaceKitNavigation):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::createNewPage):
* Source/WebKit/UIProcess/win/AutomationSessionClientWin.cpp:
(WebKit::AutomationSessionClient::requestNewPageWithOptions):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createWindow):
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didSameDocumentNavigationForFrameViaJS):
(WebKit::WebLocalFrameLoaderClient::dispatchDecidePolicyForNewWindowAction):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):
* Source/WebKitLegacy/ios/WebCoreSupport/WebFixedPositionContent.mm:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, LocalIframeOpensBlobURLFromFileMainFrame)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/blob-popup-file-mainframe.html: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/blob-popup-local-iframe.html: Added.

Canonical link: https://commits.webkit.org/303469@main
